File: go/indicators/common/absolutepriceoscillator/absolutepriceoscillator.go
package absolutepriceoscillator

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// AbsolutePriceOscillator is the Absolute Price Oscillator (APO).
//
// APO is calculated by subtracting the slower moving average from the faster moving average.
//
//	APO = fast_ma - slow_ma
type AbsolutePriceOscillator struct {
	mu sync.RWMutex
	core.LineIndicator
	fastMA lineUpdater
	slowMA lineUpdater
	value  float64
	primed bool
}

// NewAbsolutePriceOscillator returns an instance of the indicator created using supplied parameters.
func NewAbsolutePriceOscillator(p *AbsolutePriceOscillatorParams) (*AbsolutePriceOscillator, error) {
	const (
		invalid   = "invalid absolute price oscillator parameters"
		fmts      = "%s: %s"
		fmtw      = "%s: %w"
		minLength = 2
	)

	if p.FastLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "fast length should be greater than 1")
	}

	if p.SlowLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow length should be greater than 1")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	var fastMA, slowMA lineUpdater

	var maLabel string

	switch p.MovingAverageType {
	case EMA:
		maLabel = "EMA"

		fast, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         p.FastLength,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		slow, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         p.SlowLength,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		fastMA = fast
		slowMA = slow
	default:
		maLabel = "SMA"

		fast, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: p.FastLength})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		slow, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: p.SlowLength})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		fastMA = fast
		slowMA = slow
	}

	mnemonic := fmt.Sprintf("apo(%s%d/%s%d%s)", maLabel, p.FastLength, maLabel, p.SlowLength,
		core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Absolute Price Oscillator " + mnemonic

	apo := &AbsolutePriceOscillator{
		fastMA: fastMA,
		slowMA: slowMA,
		value:  math.NaN(),
	}

	apo.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, apo.Update)

	return apo, nil
}

// IsPrimed indicates whether the indicator is primed.
func (a *AbsolutePriceOscillator) IsPrimed() bool {
	a.mu.RLock()
	defer a.mu.RUnlock()

	return a.primed
}

// Metadata describes the output data of the indicator.
func (a *AbsolutePriceOscillator) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.AbsolutePriceOscillator,
		Mnemonic:    a.LineIndicator.Mnemonic,
		Description: a.LineIndicator.Description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(AbsolutePriceOscillatorValue),
				Type:        outputs.ScalarType,
				Mnemonic:    a.LineIndicator.Mnemonic,
				Description: a.LineIndicator.Description,
			},
		},
	}
}

// Update updates the value of the indicator given the next sample.
func (a *AbsolutePriceOscillator) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	a.mu.Lock()
	defer a.mu.Unlock()

	slow := a.slowMA.Update(sample)
	fast := a.fastMA.Update(sample)
	a.primed = a.slowMA.IsPrimed() && a.fastMA.IsPrimed()

	if math.IsNaN(fast) || math.IsNaN(slow) {
		a.value = math.NaN()

		return a.value
	}

	a.value = fast - slow

	return a.value
}

File: ts/indicators/common/absolute-price-oscillator/absolute-price-oscillator.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { ExponentialMovingAverage } from '../../common/exponential-moving-average/exponential-moving-average';
import { SimpleMovingAverage } from '../../common/simple-moving-average/simple-moving-average';
import { AbsolutePriceOscillatorOutput } from './absolute-price-oscillator-output';
import { AbsolutePriceOscillatorParams, MovingAverageType } from './absolute-price-oscillator-params';

/** Interface for an indicator that accepts a scalar and returns a value. */
interface LineUpdater {
  update(sample: number): number;
  isPrimed(): boolean;
}

/** Function to calculate mnemonic of an __AbsolutePriceOscillator__ indicator. */
export const absolutePriceOscillatorMnemonic = (params: AbsolutePriceOscillatorParams): string => {
  const cm = componentTripleMnemonic(
    params.barComponent,
    params.quoteComponent,
    params.tradeComponent,
  );

  const maLabel = params.movingAverageType === MovingAverageType.EMA ? 'EMA' : 'SMA';

  return `apo(${maLabel}${params.fastLength}/${maLabel}${params.slowLength}${cm})`;
};

/**
 * AbsolutePriceOscillator is the Absolute Price Oscillator (APO).
 *
 * APO is calculated by subtracting the slower moving average from the faster moving average.
 *
 *   APO = fast_ma - slow_ma
 */
export class AbsolutePriceOscillator extends LineIndicator {
  private fastMA: LineUpdater;
  private slowMA: LineUpdater;
  private value: number;

  /**
   * Constructs an instance given the parameters.
   */
  public constructor(params: AbsolutePriceOscillatorParams) {
    super();

    const fastLength = Math.floor(params.fastLength);
    const slowLength = Math.floor(params.slowLength);

    if (fastLength < 2) {
      throw new Error('fast length should be greater than 1');
    }

    if (slowLength < 2) {
      throw new Error('slow length should be greater than 1');
    }

    this.mnemonic = absolutePriceOscillatorMnemonic(params);
    this.description = 'Absolute Price Oscillator ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;

    if (params.movingAverageType === MovingAverageType.EMA) {
      this.fastMA = new ExponentialMovingAverage({
        length: fastLength,
        firstIsAverage: params.firstIsAverage ?? false,
      });
      this.slowMA = new ExponentialMovingAverage({
        length: slowLength,
        firstIsAverage: params.firstIsAverage ?? false,
      });
    } else {
      this.fastMA = new SimpleMovingAverage({ length: fastLength });
      this.slowMA = new SimpleMovingAverage({ length: slowLength });
    }

    this.value = Number.NaN;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.AbsolutePriceOscillator,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: AbsolutePriceOscillatorOutput.AbsolutePriceOscillatorValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    const slow = this.slowMA.update(sample);
    const fast = this.fastMA.update(sample);
    this.primed = this.slowMA.isPrimed() && this.fastMA.isPrimed();

    if (Number.isNaN(fast) || Number.isNaN(slow)) {
      this.value = Number.NaN;
      return this.value;
    }

    this.value = fast - slow;

    return this.value;
  }
}
